fahrschein-opentelemetry/src/main/java/org/zalando/fahrschein/opentelemetry/InstrumentedPublishingHandler.java

```diff
@@ -0,0 +1,56 @@
+package org.zalando.fahrschein.opentelemetry;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import org.zalando.fahrschein.EventPublishingHandler;
+
+import java.util.List;
+import java.util.Locale;
+
+import static io.opentelemetry.api.trace.SpanKind.PRODUCER;
+
+/**
+ * Instruments publishing requests to Nakadi by making use of OpenTelemetry.
+ */
+public class InstrumentedPublishingHandler implements EventPublishingHandler {
+
+    private final Tracer tracer;
+
+    public InstrumentedPublishingHandler(Tracer tracer) {
+        this.tracer = tracer;
+    }
+
+    @Override
+    public <T> void onPublish(String eventName, List<T> events) {
+        Span span = tracer
+                .spanBuilder("send_" + eventName)
+                .setParent(Context.current())
+                .setSpanKind(PRODUCER)
+                .setAttribute("messaging.destination_kind", "topic")
+                .setAttribute("messaging.destination", eventName)
+                .setAttribute("messaging.system", "Nakadi")
+                .setAttribute("messaging.message_payload_size", sizeBucket(events.size()))
+                .startSpan();
+        span.makeCurrent();
+    }
+
+    public void onSuccess() {
+        Span.current().end();
+    }
+
+    @Override
+    public <T> void onError(List<T> successEvents, List<T> failedEvents, Throwable t) {
```

Comment:
I thought we were going to go with a heterogeneous list of event containers denoting success/failure, thereby preserving ordering. This was suggested by @otrosien. I think it's a lovely idea.

Reply:
yep. Let's say you have in one batch two events for the same compaction key, the first one fails the second one succeeds, you shouldn't retry the second one. Given two lists it will be impossible to reconstruct the ordering.